scripts/Diffraction/isis_powder/routines/focus.py

```diff
@@ -151,18 +156,106 @@ def _focus_one_ws(
+        if summed_empty:
+            input_workspace = common.subtract_summed_runs(ws_to_correct=input_workspace, empty_ws=summed_empty)
+        # Crop to largest acceptable TOF range
+        input_workspace = instrument._crop_raw_to_expected_tof_range(ws_to_crop=input_workspace)
+        if absorb:
+            input_workspace = instrument._apply_absorb_corrections(run_details=run_details, ws_to_correct=input_workspace)
+        else:
+            # Set sample material if specified by the user
+            if sample_details is not None:
+                mantid.SetSample(
+                    InputWorkspace=input_workspace,
+                    Geometry=sample_details.generate_sample_geometry(),
+                    Material=sample_details.generate_sample_material(),
+                )
+    return input_workspace
+
+
+def apply_per_detector_corrections(input_workspace, instrument, perform_vanadium_norm, vanadium_ws, sample_details, run_details):
+    # apply per detector vanadium correction on uncalibrated data
+    input_workspace = _apply_vanadium_corrections_per_detector(
+        instrument=instrument,
+        input_workspace=input_workspace,
+        perform_vanadium_norm=perform_vanadium_norm,
+        vanadium_splines=vanadium_ws,
+        run_details=run_details,
+    )
+    input_workspace = instrument.apply_additional_per_detector_corrections(input_workspace, sample_details, run_details)
+    return input_workspace
+
+
+def _apply_vanadium_corrections(instrument, input_workspace, vanadium_splines):
     split_data_spectra = common.extract_ws_spectra(input_workspace)
 
-    if perform_vanadium_norm:
-        processed_spectra = _normalize_spectra(spectra_list=split_data_spectra, vanadium_splines=vanadium_splines, instrument=instrument)
-    else:
-        processed_spectra = split_data_spectra
+    processed_spectra = _normalize_spectra(spectra_list=split_data_spectra, vanadium_splines=vanadium_splines, instrument=instrument)
 
     return processed_spectra
 
 
+def _apply_vanadium_corrections_per_detector(
+    instrument, input_workspace: Workspace2D, perform_vanadium_norm, vanadium_splines: Workspace2D, run_details
+):
+    input_workspace = mantid.ConvertUnits(InputWorkspace=input_workspace, OutputWorkspace=input_workspace, Target="TOF")
+    if perform_vanadium_norm:
+        input_workspace, vanadium_splines = common._remove_masked_and_monitor_spectra(
+            data_workspace=input_workspace, correction_workspace=vanadium_splines, run_details=run_details
+        )
+        processed_workspace = _normalize_per_detector_workspace(input_workspace, vanadium_splines, instrument)
+        processed_workspace = mantid.ReplaceSpecialValues(
+            InputWorkspace=processed_workspace, OutputWorkspace=processed_workspace, NaNValue=0, InfinityValue=0
+        )
+    else:
+        processed_workspace = input_workspace
+    return processed_workspace
+
+
+def divide_by_number_of_detectors_in_bank(focussed_data, cal_filepath):
```

Comment:
Is this a (near?) duplicate of what is defined in the algorithm `TotScatCalculateSelfScattering`? could be defined there and imported here?